### ISSUE TITLE
SerializableClosure::setSecretKey should be called outside of deferred ServiceProvider

### ISSUE THREAD

**Issue by ArondeParon** [laravel/framework#28005]
- Laravel Version: 5.8.4
- PHP Version: 7.3.3

### Description:

[`Illuminate\Queue\QueueServiceProvider`](https://github.com/laravel/framework/blob/5.8/src/Illuminate/Queue/QueueServiceProvider.php#L33) registers a secret key that is used to encode `Closure` objects when serializing queued objects.

The problem, however, is that the `QueueServiceProvider` is a deferred serviceprovider, and that there are no guarantees that the `queue` container object will be resolved at the moment of unserializing. When this happens, a user would see the following error:

> The serialized closure is not signed. Make sure you use a security provider for both serialization and unserializing.

Since `SerializableClosure` is not a dependency of the `Queue` per sé, it would probably be better to register this secret outside of the `QueueServiceProvider` so that it is guaranteed that this secret exists.

Related issue: https://github.com/spatie/async/issues/52#issuecomment-476263636

**Comment by driesvints** [comment]
Hmm, I can indeed see this being a problem although I'm a bit torn between it being a bug or bad usage. Note that `SerializableClosure` isn't documented and mostly is an internal class.

I saw you noted:

> Bonus: An extra complication that we experienced was the fact that things did work when setting APP_DEBUG to true. This actually helped us to find the problem: in our case, it was Laravel Debugbar that called the mail container, which in turn calls (and initialized) the queue container.

Does that mean that this can be solved within the Laravel Debugbar package?

**Comment by ArondeParon** [comment]
Hi Dries, the problem does not exactly lie within the Laravel Debugbar, but it has the side effect of accessing the `queue` from the container, causing it to be initialized. This could in theory lead to differences between `local` and `production`.

The experiences the actual issue with `Maatwebsite/Excel` when creating a queued export. This package also relies on `SerializableClosure` and caused issues when unserializing before the queue was initialized.

**Comment by driesvints** [comment]
I see. Thanks for explaining. In that case I believe it's indeed best to look for a different approach to this. One thing that comes to mind is a config file option with an env variable.

**Comment by msarca** [comment]
@ArondeParon The spatie/async#52 issue is not really related to this. The way spatie/async works is by launching a new process using the following command:

```bash
exec php /path/to/../ChildRuntime.php /path/to/../autoload.php serialized_closure
```
If you take a look at [ChildRuntime.php](/spatie/async/blob/master/src/Runtime/ChildRuntime.php) you will see that there are no calls to `SerializableClosure::setSecretKey` or `SerializableClosure::addSecurityProvider`, which means that there is no support for cryptographically signed closures. So if the `serialized_closure` argument is a signed closure, there would be an error. Since the `serialized_closure` is not sent over the network, but used as an argument to a locally executed command, is not really necessary to be signed. The [solution](/spatie/async/issues/52#issuecomment-443354354) proposed to fix this for spatie/async was not yet implemented by maintainers.

**Comment by patrickbrouwers** [comment]
Sorry to bump up an old issue

@ArondeParon I might have found a "fix" for the issue with Laravel Excel. I'm not really sure why one time the queue provider is resolved and the other time not. But I've added a hack in there to ensure it's always loaded when we start serializing. If you want you could test the fix with this PR: https://github.com/Maatwebsite/Laravel-Excel/pull/2257

@driesvints Would it somehow be possible to load this more elegantly? Perhaps a service we could call that sets up the security provider? I'm okay with the current workaround, but afraid this might break in the future?

**Comment by driesvints** [comment]
@patrickbrouwers yes that seems reasonable. Maybe attempt a PR?

**Comment by taylorotwell** [comment]
Fixed.

**Comment by deevus** [comment]
How was this fixed? I'm seeing this issue on v6.20.16.

**Comment by deevus** [comment]
Ok I think that the issue is that we were using `Opis/Closure` separately and I did not realise that Laravel added a serialized closure feature. We introduced it around Laravel 5.4, and looking at the commits the official solution came with 5.7
